freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/SyncOperationStatusService.java

```diff
@@ -67,15 +67,29 @@ public SyncOperation startOperation(String accountId, SyncOperationType syncOper
         return syncOperationRepository.save(syncOperation);
     }
 
-    public SyncOperation completeOperation(String operationId, List<SuccessDetails> success, List<FailureDetails> failure) {
+    public SyncOperation completeOperation(String operationId) {
         SyncOperation syncOperation = getSyncOperationForOperationId(operationId);
         syncOperation.setStatus(SynchronizationStatus.COMPLETED);
-        syncOperation.setSuccessList(success);
-        syncOperation.setFailureList(failure);
         syncOperation.setEndTime(System.currentTimeMillis());
         return syncOperationRepository.save(syncOperation);
     }
 
+    public SyncOperation updateOperation(String operationId, SuccessDetails success, FailureDetails failure) {
+        // TODO: Make this as fire and forget
+        SyncOperation syncOperation = getSyncOperationForOperationId(operationId);
+        if (success != null) {
+            List<SuccessDetails> successList = syncOperation.getSuccessList();
+            successList.add(success);
+            syncOperation.setSuccessList(successList);
+        }
+        if (failure != null) {
+            List<FailureDetails> failureList = syncOperation.getFailureList();
+            failureList.add(failure);
+            syncOperation.setFailureList(failureList);
+        }
+        return syncOperationRepository.save(syncOperation);
```

Comment:
you need to catch and retry on OptimisticLockingExceptions. you also need to handle if the update operation method fails (i.e., out of retries).